nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java

```diff
@@ -0,0 +1,225 @@
+                                trustStore.load(is, sslService.getTrustStorePassword().toCharArray());
+                            }
+
+                            final KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory
+                                    .getDefaultAlgorithm());
+                            kmf.init(keyStore, sslService.getKeyStorePassword().toCharArray());
+                            final TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory
+                                    .getDefaultAlgorithm());
+                            tmf.init(keyStore);
+                            SSLContext context1 = SSLContext.getInstance(sslService.getSslAlgorithm());
+                            context1.init(kmf.getKeyManagers(), tmf.getTrustManagers(), new SecureRandom());
+                            httpClientBuilder = httpClientBuilder.setSSLContext(context1);
+                        } catch (Exception e) {
+                            getLogger().error("Error setting up SSL.", e);
+                        }
+                    }
+
+                    if (username != null && password != null) {
+                        final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+                        credentialsProvider.setCredentials(AuthScope.ANY,
+                                new UsernamePasswordCredentials(username, password));
+                        httpClientBuilder = httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+                    }
+
+                    return httpClientBuilder;
+                })
+                .setRequestConfigCallback(requestConfigBuilder -> {
+                    requestConfigBuilder.setConnectTimeout(connectTimeout);
+                    requestConfigBuilder.setSocketTimeout(readTimeout);
+                    return requestConfigBuilder;
+                })
+                .setMaxRetryTimeoutMillis(retryTimeout);
+
+        this.client = builder.build();
+    }
+
+    private Response runQuery(String query, String index, String type) throws IOException {
+        StringBuilder sb = new StringBuilder()
+                .append("/")
+                .append(index);
+        if (type != null && !type.equals("")) {
+            sb.append("/")
+                    .append(type);
+        }
+
+        sb.append("/_search");
+
+        HttpEntity queryEntity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+
+        return client.performRequest("POST", sb.toString(), Collections.emptyMap(), queryEntity);
+    }
+
+    @Override
+    public Optional<SearchResponse> search(String query, String index, String type) throws IOException {
+        Response response = runQuery(query, index, type);
+        Map<String, Object> parsed = mapper.readValue(IOUtils.toString(response.getEntity().getContent()), Map.class);
```

Comment:
2 things:
1. IOUtils.toString is deprecated, you should pass the expected charset to use.
2. In general, you should avoid doing a lot of nesting of logic into a single line. It helps a lot with readability and diagnosing errors. For example, if we hit an NPE here, it could be for any of 3 different objects. At the very least, the IOUtils call to get the response content as a string should be on another line.